src/module.c

```diff
@@ -11628,32 +11893,34 @@ int moduleLoad(const char *path, void **module_argv, int module_argc, int is_loa
 }
 
 /* Unload the module registered with the specified name. On success
- * C_OK is returned, otherwise C_ERR is returned and errno is set
- * to the following values depending on the type of error:
- *
- * * ENONET: No such module having the specified name.
- * * EBUSY: The module exports a new data type and can only be reloaded. 
- * * EPERM: The module exports APIs which are used by other module. 
- * * EAGAIN: The module has blocked clients. 
- * * EINPROGRESS: The module holds timer not fired.
- * * ECANCELED: Unload module error.  */
-int moduleUnload(sds name) {
+ * C_OK is returned, otherwise C_ERR is returned and errmsg is set
+ * with an appropriate message. */
+int moduleUnload(sds name, const char **errmsg) {
     struct RedisModule *module = dictFetchValue(modules,name);
 
     if (module == NULL) {
-        errno = ENOENT;
+        *errmsg = "no such module with that name";
         return C_ERR;
     } else if (listLength(module->types)) {
-        errno = EBUSY;
+        *errmsg = "the module exports one or more module-side data "
+                  "types, can't unload";
         return C_ERR;
     } else if (listLength(module->usedby)) {
-        errno = EPERM;
+        *errmsg = "the module exports APIs used by other modules. "
+                  "Please unload them first and try again";
         return C_ERR;
     } else if (module->blocked_clients) {
-        errno = EAGAIN;
+        *errmsg = "the module has blocked clients. "
+                  "Please wait for them to be unblocked and try again";
         return C_ERR;
     } else if (moduleHoldsTimer(module)) {
-        errno = EINPROGRESS;
+        *errmsg = "the module holds timer that is not fired. "
+                  "Please stop the timer or wait until it fires.";
+        return C_ERR;
+    }
+    else if (moduleSearchOrUnregisterCustomAuthCBs(module, 0) && isCustomAuthInProgress()) {
```

Comment:
maybe we can just prevent modules with such a callback registered from being unloaded (like we do for modules with data types).
also, note that currently it is unsafe to even load a module that just registered commands (no data types of callbacks / hooks), see #10177. so module unloading is not really properly supported, and i think we can just give up that effort and forbid it.

p.s. styling thing about `} else`

Reply:
I'm generally okay with module unloading being more of a "testing thing", and not something we really expect people to use much in production. IIRC, the original motivation of unload was to make it easier to rapidly test functionality, not that is was being used much in production.

Reply:
We are removing all new logic added to Module Unload. It will behave exactly the same as it does currently.

If there is any client that is blocked on the Module (`module->blocked_clients`), then Module unload is prevented.

This logic exists today in the moduleUnload fn
```
    } else if (module->blocked_clients) {
        *errmsg = "the module has blocked clients. "
                  "Please wait for them to be unblocked and try again";
        return C_ERR;
    } 
```